src/Controller/Controller.php

```diff
@@ -757,12 +758,71 @@ public function render(?string $template = null, ?string $layout = null): Respon
         if ($builder->getTemplate() === null) {
             $builder->setTemplate($this->request->getParam('action'));
         }
+        $viewClass = $this->chooseViewClass();
+        $view = $this->createView($viewClass);
 
-        $view = $this->createView();
         $contents = $view->render();
-        $this->setResponse($view->getResponse()->withStringBody($contents));
+        $response = $view->getResponse()->withStringBody($contents);
 
-        return $this->response;
+        return $this->setResponse($response)->response;
+    }
+
+    /**
+     * Get the list of View classes this controller can negotiate with.
+     *
+     * Each view class must implement the `getContentType()` hook method
+     * to participate in negotiation.
+     *
+     * @see Cake\Http\ContentTypeNegotiation
+     * @return array<string>
+     */
+    public function viewClasses(): array
+    {
+        return [];
+    }
+
+    /**
+     * Use the view classes defined on this controller to view
+     * selection based on content-type negotiation.
+     *
+     * @return string|null The chosen view class or null for no decision.
+     */
+    protected function chooseViewClass(): ?string
+    {
+        $possibleViewClasses = $this->viewClasses();
+        if (empty($possibleViewClasses)) {
+            return null;
+        }
+        if ($this->viewBuilder()->getClassName() !== null) {
+            return null;
```

Comment:
Why does this return null?

Reply:
Either the Controller's action, `beforeRender()` logic or a component set the viewClass so we shouldn't overwrite that decision with content-type negotiation. I've added a comment.

Reply:
Ah, I misread the check as === null.